File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    void solve(int i, vector<int> &nums, vector<vector<int>> &res){
        unordered_set<int> seen;
        for(int j = i + 1; j < nums.size(); j++){
            int complement = -nums[i] - nums[j];
            if(seen.count(complement)){
                res.push_back({nums[i], nums[j], complement});
                while(j + 1 < nums.size() && nums[j] == nums[j + 1]){
                    j++;
                }
            }
            seen.insert(nums[j]);
        }
    }
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for(int i = 0; i < nums.size() && nums[i] <= 0; i++){
            if(i == 0 || nums[i] != nums[i - 1]){
                solve(i, nums, res);
            }
        }
        return res;
    }
};
/*
// 1 last try

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;        
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] > 0) {
                break;
            }
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            } 
            int left = i + 1, right = nums.size() - 1;
            while (left < right) {
                int three_sum = nums[i] + nums[left] + nums[right];
                if (three_sum < 0) {
                    left++;
                } else if (three_sum > 0) {
                    right--;
                } else {
                    result.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                }
            }
        }
        return result;
    }
};

// Time: O(N^2)
// Space: O(1) ~ O(N)
*/

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        for i, n in enumerate(nums):
            if n > 0:
                break
            if i > 0 and n == nums[i - 1]:
                continue
            
            l, r = i + 1, len(nums) - 1
            while l < r:
                sum = n + nums[l] + nums[r]
                if sum > 0:
                    r -= 1
                elif sum < 0:
                    l += 1
                else:
                    result.append([n, nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
        return result
    
'''
Time complexity: O(n^2)
Space Complexity: In this example O(1), as the sort method sorts the list in place, meaning it doesn't create a new list. It modifies the original list.
But it may be from O(log⁡n) to O(n), depending on the implementation of the sorting algorithm. For the purpose of complexity analysis, we ignore the memory required for the output.
'''